File: cs/src/AutoDiff/Product.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace AutoDiff
{
    /// <summary>
    /// Represents a product between two terms.
    /// </summary>
    [Serializable]
    public class Product : Term
    {
        /// <summary>
        /// Constructs a new instance of the <see cref="Product"/> type.
        /// </summary>
        /// <param name="left">The first product term</param>
        /// <param name="right">The second product term</param>
        public Product(Term left, Term right)
        {
            Left = left;
            Right = right;
        }

        /// <summary>
        /// Gets the first product term.
        /// </summary>
        public Term Left { get; private set; }

        /// <summary>
        /// Gets the second product term.
        /// </summary>
        public Term Right { get; private set; }

        /// <summary>
        /// Accepts a term visitor
        /// </summary>
        /// <param name="visitor">The term visitor to accept</param>
        public override void Accept(ITermVisitor visitor)
        {
            visitor.Visit(this);
        }

        /// <summary>
        /// Accepts a term visitor with a generic result
        /// </summary>
        /// <typeparam name="TResult">The type of the result from the visitor's function</typeparam>
        /// <param name="visitor">The visitor to accept</param>
        /// <returns>
        /// The result from the visitor's visit function.
        /// </returns>
        public override TResult Accept<TResult>(ITermVisitor<TResult> visitor)
        {
            return visitor.Visit(this);
        }
		
		public override string ToString()
		{
			return string.Format("( {0} * {1} )", Left, Right);
		}
		
		public override Term AggregateConstants()
		{
			Left = Left.AggregateConstants();
			Right = Right.AggregateConstants();
			if (Left is Constant && Right is Constant) {
				return (Left as Constant).Value * (Right as Constant).Value;
			} else if (Left is Zero) {
				return Left;
			} else if (Right is Zero) {
				return Right;			
			} 
			if (Left is Constant) {
				if (((Constant)Left).Value == 1) {
					return Right;
				}
			}
			if (Right is Constant) {
				if (((Constant)Right).Value == 1) {
					return Left;
				}			
			}
			return this;			
		}
		public override Term Derivative(Variable v)
		{
			return this.Left*this.Right.Derivative(v) + this.Right*this.Left.Derivative(v);
		}
    }
}

File: java/src/main/java/AutoDiff/Product.java
//
// Translated by CS2J (http://www.cs2j.com): 12.11.2016 09:40:34
//

package AutoDiff;

/**
* Represents a product between two terms.
*/
public class Product  extends Term 
{
    /**
    * Constructs a new instance of the 
    *  {@link #Product}
    *  type.
    * 
    *  @param left The first product term
    *  @param right The second product term
    */
    public Product(Term left, Term right) throws Exception {
        setLeft(left);
        setRight(right);
    }

    /**
    * Gets the first product term.
    */
    private Term Left;
    public Term getLeft() {
        return Left;
    }

    public void setLeft(Term value) {
        Left = value;
    }

    /**
    * Gets the second product term.
    */
    private Term Right;
    public Term getRight() {
        return Right;
    }

    public void setRight(Term value) {
        Right = value;
    }

//    /**
//    * Accepts a term visitor
//    *
//    *  @param visitor The term visitor to accept
//    */
//    public void accept(ITermVisitor visitor) throws Exception {
//        visitor.visit(this);
//    }

    /**
    * Accepts a term visitor with a generic result
    * The type of the result from the visitor's function
    *  @param visitor The visitor to accept
    *  @return 
    * The result from the visitor's visit function.
    */
    public <TResult>TResult accept(ITermVisitor<TResult> visitor) throws Exception {
        return visitor.visit(this);
    }

    public String toString() {
        try
        {
            return String.format("( %s * %s )", getLeft(), getRight());
        }
        catch (RuntimeException exception)
        {
            throw exception;
        }
        catch (Exception exception)
        {
            throw new RuntimeException(exception);
        }
    
    }

    public Term aggregateConstants() throws Exception {
        setLeft(getLeft().aggregateConstants());
        setRight(getRight().aggregateConstants());
        if (getLeft() instanceof Constant && getRight() instanceof Constant)
        {
            return TermBuilder.Constant((getLeft() instanceof Constant ? (Constant)getLeft() : (Constant)null)
                .getValue() *
                (getRight() instanceof Constant ? (Constant)getRight() : (Constant)null).getValue());
        }
        else if (getLeft() instanceof Zero)
        {
            return getLeft();
        }
        else if (getRight() instanceof Zero)
        {
            return getRight();
        }
           
        if (getLeft() instanceof Constant)
        {
            if (((Constant)getLeft()).getValue() == 1)
            {
                return getRight();
            }
             
        }
         
        if (getRight() instanceof Constant)
        {
            if (((Constant)getRight()).getValue() == 1)
            {
                return getLeft();
            }
             
        }
         
        return this;
    }

    public Term derivative(Variable v) throws Exception {
        return add(multiply(this.getLeft(), this.getRight().derivative(v)), multiply(this.getRight(),
            this.getLeft().derivative(v)));
    }

}


